megamek/src/megamek/client/bot/BotClient.java

```diff
@@ -915,33 +920,30 @@ protected Coords getFirstValidCoords(Entity deployedUnit, List<Coords> possibleD
         return result;
     }
 
-    /**
-     * Gets the {@link BoardEdgePathFinder} for an {@link Entity} for their
-     * current movement mode.
-     * @param entity The entity to retrieve the {@link BoardEdgePathFinder}.
-     * @return The appropriate {@link BoardEdgePathFinder} for the given entity.
-     */
-    private BoardEdgePathFinder getBoardEdgePathFinder(Entity entity) {
-        return deploymentPathFinders.computeIfAbsent(entity.getMovementMode(), 
-            e -> new BoardEdgePathFinder());
-    }
-
-    // ToDo: Change this to 'hasSafePathToCenter' to account for buildings, lava and similar hazards.
     /**
      * Determines if the given entity has a reasonable path to the "opposite" edge of the board from its
-     * current position.
-     * @param entity
-     * @param board
-     * @return
+     * current position. Returns 0 if this can be accomplished without destroying any terrain, 
+     * -50 if this can be accomplished but terrain must be destroyed,
+     * -100 if this cannot be accomplished at all
      */
-    private boolean hasPathToEdge(Entity entity, IBoard board) {
+    private int hasPathToEdge(Entity entity, IBoard board) {
         // Flying units can always get anywhere
         if (entity.isAirborne() || entity instanceof VTOL) {
-            return true;
+            return 0;
         }
         
-        MovePath mp = getBoardEdgePathFinder(entity).findPathToEdge(entity);
-        return mp != null;
+        CardinalEdge destinationEdge = BoardUtilities.determineOppositeEdge(entity);
+        
+        int noReductionZoneSize = getClusterTracker().getDestinationCoords(entity, destinationEdge, false).size();
+        int reductionZoneSize = getClusterTracker().getDestinationCoords(entity, destinationEdge, true).size();
```

Comment:
If this is expensive you could move this into an else block after the `noReductionZoneSize > 0` check.

Reply:
Running time of n hashmap lookups where n is the width/height of the edge, so probably not worth it.